runners/spark/src/main/java/org/apache/beam/runners/spark/translation/ValueAndCoderKryoSerializable.java

```diff
@@ -0,0 +1,203 @@
+ * instance via getOrDecode(Coder). This reverts the representation back to the deserialized
+ * representation.
+ *
+ * @param <T> element type
+ */
+public class ValueAndCoderKryoSerializable<T> implements KryoSerializable, Externalizable {
+  private T value;
+  // Re-use a field to save space in-memory. This is either a byte[] or a Coder, depending on
+  // which representation we are in.
+  private Object coderOrBytes;
+
+  ValueAndCoderKryoSerializable(T value, Coder<T> currentCoder) {
+    this.value = value;
+    this.coderOrBytes = currentCoder;
+  }
+
+  @SuppressWarnings("unused") // for serialization
+  public ValueAndCoderKryoSerializable() {}
+
+  public T getOrDecode(Coder<T> coder) throws IOException {
+    if (!(coderOrBytes instanceof Coder)) {
+      value =
+          coder.decode(new ByteArrayInputStream((byte[]) this.coderOrBytes), Coder.Context.OUTER);
+      this.coderOrBytes = coder;
+    }
+
+    return value;
+  }
+
+  private static class ByteSizeObserver extends ElementByteSizeObserver {
+    private long observedSize = 0;
+
+    @Override
+    protected void reportElementSize(long elementByteSize) {
+      observedSize += elementByteSize;
+    }
+  }
+
+  private void writeCommon(OutputStream out) throws IOException {
+    if (!(coderOrBytes instanceof Coder)) {
+      byte[] bytes = (byte[]) coderOrBytes;
+      VarInt.encode(bytes.length, out);
+      out.write(bytes);
+    } else {
+      @SuppressWarnings("unchecked")
+      Coder<T> coder = (Coder<T>) coderOrBytes;
+      int bufferSize = 1024;
+
+      if (coder.isRegisterByteSizeObserverCheap(value)) {
+        try {
+          ByteSizeObserver observer = new ByteSizeObserver();
+          coder.registerByteSizeObserver(value, observer);
+          bufferSize = (int) observer.observedSize;
+        } catch (Exception e) {
+          Throwables.throwIfUnchecked(e);
+          throw new RuntimeException(e);
+        }
+      }
+
+      ByteArrayOutputStream bytes = new ByteArrayOutputStream(bufferSize);
+
+      try {
+        coder.encode(value, bytes, Coder.Context.OUTER);
+
+        VarInt.encode(bytes.size(), out);
+        bytes.writeTo(out);
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  private void readCommon(InputStream in) throws IOException {
+    int length = VarInt.decodeInt(in);
+    byte[] bytes = new byte[length];
+    ByteStreams.readFully(in, bytes);
+    this.coderOrBytes = bytes;
+  }
+
+  @Override
+  public void write(Kryo kryo, Output output) {
+    try {
+      writeCommon(output);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public void read(Kryo kryo, Input input) {
+    try {
+      readCommon(input);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public void writeExternal(final ObjectOutput out) throws IOException {
```

Comment:
Is this better in some sense than just writing `writeObject()` as stated in [Serializable documentation](https://docs.oracle.com/javase/8/docs/api/java/io/Serializable.html)?

Reply:
I'm not sure which is better - I switched to `writeObject`.

Reply:
Me neither but `writeObject` seems to be more commonly used.